File: 0037-sudoku-solver/0037-sudoku-solver.cpp
#include <vector>

class Solution {
private:
    bool rows[9][10] = {false};
    bool cols[9][10] = {false};
    bool boxes[9][10] = {false};

    bool backtrack(std::vector<std::vector<char>>& board, int r, int c) {
        if (r == 9) return true; // Reached the end of the board
        if (c == 9) return backtrack(board, r + 1, 0); // Move to next row
        
        if (board[r][c] != '.') {
            return backtrack(board, r, c + 1); // Skip pre-filled cells
        }

        int box_idx = (r / 3) * 3 + (c / 3);

        for (int num = 1; num <= 9; ++num) {
            if (!rows[r][num] && !cols[c][num] && !boxes[box_idx][num]) {
                // Place digit
                board[r][c] = num + '0';
                rows[r][num] = cols[c][num] = boxes[box_idx][num] = true;

                if (backtrack(board, r, c + 1)) return true;

                // Backtrack
                board[r][c] = '.';
                rows[r][num] = cols[c][num] = boxes[box_idx][num] = false;
            }
        }

        return false;
    }

public:
    void solveSudoku(std::vector<std::vector<char>>& board) {
        // Initialize tracking arrays
        for (int r = 0; r < 9; ++r) {
            for (int c = 0; c < 9; ++c) {
                if (board[r][c] != '.') {
                    int num = board[r][c] - '0';
                    int box_idx = (r / 3) * 3 + (c / 3);
                    rows[r][num] = cols[c][num] = boxes[box_idx][num] = true;
                }
            }
        }
        backtrack(board, 0, 0);
    }
};

File: 0037-sudoku-solver/0037-sudoku-solver.py
class Solution:

  def solveSudoku(self, board: list[list[str]]) -> None:
    """Modifies board in-place to solve the Sudoku puzzle."""
    rows = [set() for _ in range(9)]
    cols = [set() for _ in range(9)]
    boxes = [set() for _ in range(9)]
    empty_cells = []

    # Step 1: Pre-fill constraints and locate empty cells
    for r in range(9):
      for c in range(9):
        val = board[r][c]
        if val != ".":
          rows[r].add(val)
          cols[c].add(val)
          boxes[(r // 3) * 3 + (c // 3)].add(val)
        else:
          empty_cells.append((r, c))

    # Step 2: Backtracking helper function
    def backtrack(idx: int) -> bool:
      if idx == len(empty_cells):
        return True  # All empty cells filled successfully

      r, c = empty_cells[idx]
      box_idx = (r // 3) * 3 + (c // 3)

      for digit in "123456789":
        if (
            digit not in rows[r]
            and digit not in cols[c]
            and digit not in boxes[box_idx]
        ):
          # Make move
          board[r][c] = digit
          rows[r].add(digit)
          cols[c].add(digit)
          boxes[box_idx].add(digit)

          if backtrack(idx + 1):
            return True

          # Backtrack (Undo move)
          board[r][c] = "."
          rows[r].remove(digit)
          cols[c].remove(digit)
          boxes[box_idx].remove(digit)

      return False

    backtrack(0)